### ISSUE TITLE
Assertion failed: sched::exception_depth <= 1 (core/sched.cc: reschedule_from_interrupt: 235)

### ISSUE THREAD

**Issue by justinc1** [cloudius-systems/osv#711]
When running hello-world mpi program, with many more workers then actual cpu cores, assertion fails about 4-times in 100 runs. Command to run (just as reminder to me, for later patch testing) - "mpirun -n 46 -H 192.168.122.11 --launch-agent /home/justin_cinkelj/devel/mikelangelo/osv_proxy/lin_proxy.sh -wd /  /usr/lib/mpi_hello.so 192.168.122.1 8080 0".

Backtrace has two variants:
```
Assertion failed: sched::exception_depth <= 1 (core/sched.cc: reschedule_from_interrupt: 235)

[backtrace]
0x000000000022bc6d <abort(char const*, ...)+249>
0x000000000022bcce <__assert_fail+58>
0x00000000005b956f <sched::cpu::reschedule_from_interrupt(bool, std::chrono::duration<long, std::ratio<1l, 1000000000l> >)+83>
0x00000000005b94eb <sched::cpu::schedule()+71>
0x00000000005bc8c6 <sched::thread::wait()+26>
0x0000000000444f32 <void sched::thread::do_wait_until<sched::noninterruptible, sched::thread::dummy_lock, waiter::wait() const::{lambda()#1}>(sched::thread::dummy_lock&, waiter::wait() const::{lambda()#1})+190>
0x0000000000443bca <void sched::thread::wait_until<waiter::wait() const::{lambda()#1}>(waiter::wait() const::{lambda()#1})+30>
```

and

```
Assertion failed: sched::exception_depth <= 1 (core/sched.cc: reschedule_from_interrupt: 235)

[backtrace]
0x000000000022bc6d <abort(char const*, ...)+249>
0x000000000022bcce <__assert_fail+58>
0x00000000005b956f <sched::cpu::reschedule_from_interrupt(bool, std::chrono::duration<long, std::ratio<1l, 1000000000l> >)+83>
0x00000000005b94eb <sched::cpu::schedule()+71>
0x00000000005bc8c6 <sched::thread::wait()+26>
0x0000000000612202 <void sched::thread::do_wait_for<lockfree::mutex, sched::wait_object<waitqueue> >(lockfree::mutex&, sched::wait_object<waitqueue>&&)+124>
0x0000000000612183 <void sched::thread::wait_for<waitqueue&>(lockfree::mutex&, waitqueue&)+57>
0x000000000061140b <waitqueue::wait(lockfree::mutex&)+51>
```

**Comment by nyh** [comment]
Can you please try to connect gdb to one of these crashes and provide a backtrace from it? I'm hoping to see a bit more context on where this happens, I suspect the real problem (which led us to this "nested exception" case) happened earlier in the call chain.

**Comment by justinc1** [comment]
```
(gdb) where
#0  0x000000000022bbf9 in abort (fmt=0xa2f006 "exception nested too deeply") at runtime.cc:102
#1  0x000000000048850e in sched::arch_cpu::enter_exception (this=0xffff8000019ce070) at arch/x64/arch-cpu.cc:19
#2  0x000000000048824d in sched::exception_guard::exception_guard (this=0x6fb79f <d_demangle_callback.constprop+367>) at arch/x64/arch-cpu.cc:37
#3  0x000000000048b959 in page_fault (ef=0x6fb7f8 <d_demangle_callback.constprop+456>) at arch/x64/mmu.cc:22
#4  <signal handler called>
#5  0x000000000059b64d in std::__atomic_base<memory::free_object*>::load (this=0x0, __m=std::memory_order_relaxed) at /usr/include/c++/4.8/bits/atomic_base.h:805
#6  0x00000000005997ea in std::atomic<memory::free_object*>::load (this=0x0, __m=std::memory_order_relaxed) at /usr/include/c++/4.8/atomic:386
#7  0x0000000000597905 in lockfree::unordered_queue_mpsc<memory::free_object>::push (this=0x0, item=0xffff800013ce4298) at include/lockfree/unordered-queue-mpsc.hh:48
#8  0x0000000000595dce in memory::garbage_sink::free (this=0x0, obj_cpu=0, obj=0xffff800013ce4298) at core/mempool.cc:125
#9  0x000000000058c26e in memory::pool::free_different_cpu (this=0x16c1248 <memory::malloc_pools+72>, obj=0xffff800013ce4298, obj_cpu=0, cur_cpu=27062112) at core/mempool.cc:314
#10 0x000000000058c31c in memory::pool::free (this=0x16c1248 <memory::malloc_pools+72>, object=0xffff800013ce4298) at core/mempool.cc:335
#11 0x000000000058ec98 in free (object=0xffff800013ce4298) at core/mempool.cc:1610
#12 0x00000000003fa8a9 in std::_Function_base::_Base_manager<console::console_multiplexer::write_ll(char const*, size_t)::__lambda2>::_M_destroy(std::_Any_data &, std::false_type) (__victim=...) at /usr/include/c++/4.8/functional:1926
#13 0x00000000003fa68a in std::_Function_base::_Base_manager<console::console_multiplexer::write_ll(char const*, size_t)::__lambda2>::_M_manager(std::_Any_data &, const std::_Any_data &, std::_Manager_operation) (__dest=..., 
    __source=..., __op=std::__destroy_functor) at /usr/include/c++/4.8/functional:1950
#14 0x0000000000207ea7 in std::_Function_base::~_Function_base (this=0xffff8000019ceea0, __in_chrg=<optimized out>) at /usr/include/c++/4.8/functional:2030
#15 0x00000000003fa99c in std::function<void (char const*, unsigned long)>::~function() (this=0xffff8000019ceea0, __in_chrg=<optimized out>) at /usr/include/c++/4.8/functional:2174
#16 0x00000000003f9fdb in console::console_multiplexer::write_ll (this=0x16a3b20 <console::mux>, 
    str=0x16bae60 <debug_ll::msg> "0x0000000000443bca <void sched::thread::wait_until<waiter::wait() const::{lambda()#1}>(waiter::wait() const::{lambda()#1})+30>\n", len=127) at drivers/console-multiplexer.cc:71
#17 0x00000000003f9064 in console::write_ll (msg=0x16bae60 <debug_ll::msg> "0x0000000000443bca <void sched::thread::wait_until<waiter::wait() const::{lambda()#1}>(waiter::wait() const::{lambda()#1})+30>\n", len=127)
    at drivers/console.cc:63
#18 0x000000000056dfa6 in debug_ll (fmt=0x99ec5d "0x%016lx <%s>\n") at core/debug.cc:250
#19 0x000000000022bb53 in print_backtrace () at runtime.cc:88
#20 0x000000000022bc6e in abort (fmt=0x99ec80 "Assertion failed: %s (%s: %s: %d)\n") at runtime.cc:121
#21 0x000000000022bccf in __assert_fail (expr=0xa437c1 "sched::exception_depth <= 1", file=0xa437b3 "core/sched.cc", line=235, 
    func=0xa45020 <sched::cpu::reschedule_from_interrupt(bool, std::chrono::duration<long, std::ratio<1l, 1000000000l> >)::__func__> "reschedule_from_interrupt") at runtime.cc:139
#22 0x00000000005b9570 in sched::cpu::reschedule_from_interrupt (this=0xffff8000019ce040, called_from_yield=false, preempt_after=...) at core/sched.cc:235
#23 0x00000000005b94ec in sched::cpu::schedule () at core/sched.cc:227
#24 0x00000000005bc8c7 in sched::thread::wait (this=0xffff800005159050) at core/sched.cc:1057
#25 0x0000000000444f33 in sched::thread::do_wait_until<sched::noninterruptible, sched::thread::dummy_lock, waiter::wait() const::{lambda()#1}>(sched::thread::dummy_lock&, waiter::wait() const::{lambda()#1}) (mtx=..., pred=...)
    at include/osv/sched.hh:1005
---Type <return> to continue, or q <return> to quit---
#26 0x0000000000443bcb in sched::thread::wait_until<waiter::wait() const::{lambda()#1}>(waiter::wait() const::{lambda()#1}) (pred=...) at include/osv/sched.hh:1016
#27 0x000000000044317c in waiter::wait (this=0xffff8000019cfb20) at include/osv/wait_record.hh:42
#28 0x00000000005671ed in lockfree::mutex::lock (this=0x16a3860 <mmu::vma_list_mutex>) at core/lfmutex.cc:88
#29 0x0000000000253f28 in std::lock_guard<lockfree::mutex>::lock_guard (this=0xffff8000019cfb80, __m=...) at /usr/include/c++/4.8/mutex:414
#30 0x000000000056a292 in rwlock::rlock (this=0x16a3860 <mmu::vma_list_mutex>) at core/rwlock.cc:28
#31 0x00000000003ccc80 in rwlock_for_read::lock (this=0x16a3860 <mmu::vma_list_mutex>) at include/osv/rwlock.h:113
#32 0x00000000003ce6b0 in std::lock_guard<rwlock_for_read&>::lock_guard (this=0xffff8000019cfc30, __m=...) at /usr/include/c++/4.8/mutex:414
#33 0x00000000003cfcc3 in lock_guard_for_with_lock<rwlock_for_read&>::lock_guard_for_with_lock (this=0xffff8000019cfc30, lock=...) at include/osv/mutex.h:75
#34 0x00000000003c66fa in mmu::vm_fault (addr=0, ef=0xffff8000019d0048) at core/mmu.cc:1334
#35 0x000000000048ba46 in page_fault (ef=0xffff8000019d0048) at arch/x64/mmu.cc:38
#36 <signal handler called>
#37 0x00000000002c8af7 in std::__atomic_base<unsigned long>::load (this=0x69, __m=std::memory_order_seq_cst) at /usr/include/c++/4.8/bits/atomic_base.h:496
#38 0x00000000004200f9 in jvm_balloon_shrinker::ballooning (this=0x1) at java/jvm/jvm_balloon.hh:36
#39 0x000000000041bbdf in memory::throttling_needed () at java/jvm/jvm_balloon.cc:85
#40 0x0000000000265a47 in vm_throttling_needed () at bsd/porting/mmu.cc:41
#41 0x000000000031b9f1 in arc_evict_needed (type=ARC_BUFC_DATA) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/arc.c:2646
#42 0x000000000031ba71 in arc_get_data_buf (buf=0xffffa00012508f00) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/arc.c:2689
#43 0x000000000031cc9c in arc_read_nolock (pio=0xffffa000084e1400, spa=0xffff9000034f7000, bp=0xffff800003d59440, done=0x323a19 <dbuf_read_done>, private=0xffffa0000c2bb600, priority=0, zio_flags=64, arc_flags=0xffff80000515cb6c, 
    zb=0xffff80000515cb70) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/arc.c:3233
#44 0x000000000031c612 in arc_read (pio=0xffffa000084e1400, spa=0xffff9000034f7000, bp=0xffff800003d59440, pbuf=0xffffa00003ab9300, done=0x323a19 <dbuf_read_done>, private=0xffffa0000c2bb600, priority=0, zio_flags=64, 
    arc_flags=0xffff80000515cb6c, zb=0xffff80000515cb70) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/arc.c:3085
#45 0x00000000003541a3 in dsl_read (pio=0xffffa000084e1400, spa=0xffff9000034f7000, bpp=0xffff800003d59440, pbuf=0xffffa00003ab9300, done=0x323a19 <dbuf_read_done>, private=0xffffa0000c2bb600, priority=0, zio_flags=64, 
    arc_flags=0xffff80000515cb6c, zb=0xffff80000515cb70) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dsl_scan.c:402
#46 0x0000000000323f2d in dbuf_read_impl (db=0xffffa0000c2bb600, zio=0xffffa000084e1400, flags=0xffff80000515cbdc) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dbuf.c:585
#47 0x000000000032413a in dbuf_read (db=0xffffa0000c2bb600, zio=0xffffa000084e1400, flags=22) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dbuf.c:633
#48 0x000000000032c469 in dmu_buf_hold_array_by_dnode_sparse (dn=0xffff900004a84000, offset=0, length=4096, read=1, tag=0xa0b23f <__func__.12821>, numbufsp=0xffff80000515cd84, dbpp=0xffff80000515cd88, flags=0, fail_sparse=1)
    at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dmu.c:427
#49 0x000000000032c716 in dmu_buf_hold_array_sparse (os=0xffff9000037ba000, object=730, offset=0, length=4096, read=1, tag=0xa0b23f <__func__.12821>, numbufsp=0xffff80000515cd84, dbpp=0xffff80000515cd88, fail_sparce=1)
    at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dmu.c:489
#50 0x000000000032d896 in dmu_map_uio (os=0xffff9000037ba000, object=730, uio=0xffff80000515ce40, size=4096) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dmu.c:1007
#51 0x00000000003a93e1 in zfs_arc (vp=0xffffa00004a8df00, fp=0xffffa00004ac0080, uio=0xffff80000515ce40) at bsd/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/zfs_vnops.c:753
---Type <return> to continue, or q <return> to quit---
#52 0x0000000000662acc in vfs_file::get_arcbuf (this=0xffffa00004ac0080, key=0xffff80000515cfc0, offset=0) at fs/vfs/vfs_fops.cc:175
#53 0x00000000005744a9 in pagecache::create_read_cached_page (fp=0xffffa00004ac0080, key=...) at core/pagecache.cc:388
#54 0x0000000000574a87 in pagecache::get (fp=0xffffa00004ac0080, offset=0, ptep=..., pte=..., write=false, shared=false) at core/pagecache.cc:470
#55 0x00000000006629c9 in vfs_file::map_page (this=0xffffa00004ac0080, off=0, ptep=..., pte=..., write=false, shared=false) at fs/vfs/vfs_fops.cc:143
#56 0x00000000003cdc5b in mmu::map_file_page_mmap::map (this=0xffffa00004a8b5e0, offset=0, ptep=..., pte=..., write=false) at core/mmu.cc:1094
#57 0x00000000003f7c37 in mmu::populate<(mmu::account_opt)1>::page<0> (this=0xffff80000515dae0, ptep=..., offset=0) at core/mmu.cc:532
#58 0x00000000003f8698 in mmu::populate_small<(mmu::account_opt)1>::page<0> (this=0xffff80000515dae0, ptep=..., offset=0) at core/mmu.cc:550
#59 0x00000000003cc0df in mmu::page<mmu::populate_small<(mmu::account_opt)1>, 0> (pops=..., ptep=..., offset=0) at core/mmu.cc:311
#60 0x00000000003f6aa0 in mmu::map_level<mmu::populate_small<(mmu::account_opt)1>, 1>::operator() (this=0xffff80000515d420, parent=..., base_virt=17592274124800) at core/mmu.cc:445
#61 0x00000000003f5103 in mmu::map_level<mmu::populate_small<(mmu::account_opt)1>, 2>::map_range<1> (this=0xffff80000515d630, vcur=17592274124800, size=4096, page_mapper=..., slop=4096, ptep=..., base_virt=17592274124800)
    at core/mmu.cc:399
#62 0x00000000003f1797 in mmu::map_level<mmu::populate_small<(mmu::account_opt)1>, 2>::operator() (this=0xffff80000515d630, parent=..., base_virt=17592274124800) at core/mmu.cc:449
#63 0x00000000003ee079 in mmu::map_level<mmu::populate_small<(mmu::account_opt)1>, 3>::map_range<2> (this=0xffff80000515d840, vcur=17592274124800, size=4096, page_mapper=..., slop=4096, ptep=..., base_virt=17592186044416)
    at core/mmu.cc:399
#64 0x00000000003e88f7 in mmu::map_level<mmu::populate_small<(mmu::account_opt)1>, 3>::operator() (this=0xffff80000515d840, parent=..., base_virt=17592186044416) at core/mmu.cc:449
#65 0x00000000003e38df in mmu::map_level<mmu::populate_small<(mmu::account_opt)1>, 4>::map_range<3> (this=0xffff80000515da50, vcur=17592274124800, size=4096, page_mapper=..., slop=4096, ptep=..., base_virt=17592186044416)
    at core/mmu.cc:399
#66 0x00000000003ddde3 in mmu::map_level<mmu::populate_small<(mmu::account_opt)1>, 4>::operator() (this=0xffff80000515da50, parent=..., base_virt=17592186044416) at core/mmu.cc:449
#67 0x00000000003d9b96 in mmu::map_range<mmu::populate_small<(mmu::account_opt)1> > (vma_start=17592274124800, vstart=17592274124800, size=4096, page_mapper=..., slop=4096) at core/mmu.cc:354
#68 0x00000000003d5103 in mmu::operate_range<mmu::populate_small<(mmu::account_opt)1> > (mapper=..., vma_start=0x100005400000, start=0x100005400000, size=4096) at core/mmu.cc:806
#69 0x00000000003d232f in mmu::vma::operate_range<mmu::populate_small<(mmu::account_opt)1> > (this=0xffffa00004ac0180, mapper=..., addr=0x100005400000, size=4096) at core/mmu.cc:1412
#70 0x00000000003cf52d in mmu::populate_vma<(mmu::account_opt)1> (vma=0xffffa00004ac0180, v=0x100005400000, size=4096, write=false) at core/mmu.cc:1208
#71 0x00000000003c7a94 in mmu::file_vma::fault (this=0xffffa00004ac0180, addr=17592274124800, ef=0xffff80000515e078) at core/mmu.cc:1702
#72 0x00000000003c67f1 in mmu::vm_fault (addr=17592274124800, ef=0xffff80000515e078) at core/mmu.cc:1341
#73 0x000000000048ba46 in page_fault (ef=0xffff80000515e078) at arch/x64/mmu.cc:38
#74 <signal handler called>
#75 0x0000000000402257 in elf::object::lookup_symbol_gnu (this=0xffffa00004ab8400, name=0x100007c0dedd "mca_oob_tcp_peer_close") at core/elf.cc:748
#76 0x000000000040248e in elf::object::lookup_symbol (this=0xffffa00004ab8400, name=0x100007c0dedd "mca_oob_tcp_peer_close") at core/elf.cc:785
#77 0x0000000000404dbb in elf::program::__lambda6::operator() (__closure=0x2000004fef50, ml=...) at core/elf.cc:1262
#78 0x0000000000406870 in elf::program::with_modules<elf::program::lookup(char const*)::__lambda6>(elf::program::__lambda6) (this=0xffff900001e0d000, f=...) at include/osv/elf.hh:641
#79 0x0000000000404e89 in elf::program::lookup (this=0xffff900001e0d000, name=0x100007c0dedd "mca_oob_tcp_peer_close") at core/elf.cc:1267
---Type <return> to continue, or q <return> to quit---
#80 0x0000000000401714 in elf::object::symbol (this=0xffffa000033c8a00, idx=321) at core/elf.cc:551
#81 0x0000000000401eb1 in elf::object::resolve_pltgot (this=0xffffa000033c8a00, index=441) at core/elf.cc:673
#82 0x0000000000405f4a in elf_resolve_pltgot (index=441, obj=0xffffa000033c8a00) at core/elf.cc:1530
#83 0x000000000048a504 in __elf_resolve_pltgot () at arch/x64/elf-dl.S:47
#84 0x000000000000ffff in ?? ()
#85 0x0000000000000000 in ?? ()

```

**Comment by nyh** [comment]
Very interesting stack trace. Here is my preliminary analysis:

It seems everything started with a symbol from a shared object that needed to be resolved ("mca_oob_tcp_peer_close"), and doing this tried to read from the mmap'ed object, which was never read before, so now we had a page fault to fetch its content.

This page fault figured out where this unread page needs to be read from, and it turns out that we (naturally) need to read it from ZFS. Reading it from ZFS means setting up a page in the ZFS "arc" cache, and later reading from disk into this cache.

Before we create a new page in the ZFS arc cache, we call arc_evict_needed() to check whether this cache is already too big and we need to start evicting old pages before creating new ones. Normally, all we need to test is whether the ARC cache's size is above its limit. But in OSv, we also added another test - we check if vm_throttling_needed(), which (as far as I understand) is supposed to tell us if some other subsystem is desperately needs more memory and was the ZFS arc cache (in this case) to start shrinking.

Our implementation of vm_throttling_needed() calls memory::throttling_needed() (see frame 39), which surprisingly is part of jvm_balloon.cc.

The throttling_needed() code is simple:

```c++
bool throttling_needed()
{
    if (!balloon_shrinker) {
        return false;
    }
    return balloon_shrinker->ballooning();
}
```
Because in this use case we did not use a JVM, or JVM ballooning, we expect balloon_shrinker to be null, so this function should do nothing. But according to frame 38, we have balloon_shrinker = 0x1, and try to call ballooning() with this = 0x1.  This is the bug.

That was the real bug, but from here obviously, all hell breaks loose: We have a second, nested, page fault, it tries to sleep and can't because nested exceptions can't sleep, this tries to print an assertion failure, and a silly bug in that (I'll open another issue on that) causes malloc() and free() to be used inside and free() crashes for the third time because it gets a bogus (see frame 9) sched::cpu::current() - I'm not exactly sure why, but it's already the third nested crash and it's not really interesting.

Anyway, the real bug was why balloon_shrinker became 0x1. Since 0x1 is not even a valid pointer, it is possible that we have some sort of memory overrun here, were some writing intended to go somewhere else and came to balloon_shrinker.
I'm not sure yet how to debug this. Maybe we can set a watchpoint for balloon_shrinker changing to anything but 0 (it is supposed to always remains 0, nothing sets it!) and figure out who writes there. I'm not sure, however, doing this will be practical considering how rare this bug is :-(

**Comment by justinc1** [comment]
Well it occurred 4-times in 100 runs. Yesterday I had to run 40 trails to get first backtrace. I guess it could be difficult/impossible to run mpi job with 100 VMs.

I will add watch on balloon_shrinker, and then post all occurrences. I can also send you app libs and OSv version/sha1 hash I'm using, if/when you need more interactive debugging.

**Comment by justinc1** [comment]
Seems balloon_shrinker gets corrupted all the time, even when there is no assert/backtrace. It was written many times, most often with a valid pointer value, and only once with value of 1.
First write during my app startup (in separate elf namespace and environ)
```
(gdb) print &balloon_shrinker
$4 = (jvm_balloon_shrinker **) 0x16a41e0 <balloon_shrinker>

Hardware watchpoint 1: *0x16a41e0

Old value = 0
New value = 222254592
elf::program::add_debugger_obj (this=0xffffa0000d2efe00, obj=0xffffa0000d3f5600) at core/elf.cc:1239
1239	}

(gdb) where
#0  elf::program::add_debugger_obj (this=0xffffa0000d2efe00, obj=0xffffa0000d3f5600) at core/elf.cc:1239
#1  0x00000000004043f5 in elf::program::load_object (this=0xffffa0000d2efe00, name="/libenviron.so", 
    extra_path=std::vector of length 0, capacity 0, loaded_objects=std::vector of length 0, capacity 0)
    at core/elf.cc:1154
#2  0x0000000000404781 in elf::program::get_library (this=0xffffa0000d2efe00, name="libenviron.so", 
    extra_path=std::vector of length 0, capacity 0) at core/elf.cc:1172
#3  0x0000000000626ddb in osv::application::clone_osv_environ (this=0xffffa0000d2b6418) at core/app.cc:409
#4  0x0000000000625a7d in osv::application::application (this=0xffffa0000d2b6418, command="/usr/lib/mpi_hello.so", 
    args=std::vector of length 4, capacity 4 = {...}, new_program=true, env=0xffffa0000d2bcc40) at core/app.cc:154
#5  0x000000000062dda7 in __gnu_cxx::new_allocator<osv::application>::construct<osv::application<std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> >
    (this=0x200003eff96f, __p=0xffffa0000d2b6418) at /usr/include/c++/4.8/ext/new_allocator.h:120
#6  0x000000000062dc52 in std::allocator_traits<std::allocator<osv::application> >::_S_construct<osv::application<std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> >(std::allocator<osv::application>&, std::allocator_traits<std::allocator<osv::application> >::__construct_helper*, (osv::application<std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&>&&)...) (__a=..., __p=0xffffa0000d2b6418)
    at /usr/include/c++/4.8/bits/alloc_traits.h:254
#7  0x000000000062db06 in std::allocator_traits<std::allocator<osv::application> >::construct<osv::application<std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> >(std::allocator<osv::application>&, osv::application<std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&>*, (osv::application<std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&>&&)...) (__a=..., __p=0xffffa0000d2b6418) at /usr/include/c++/4.8/bits/alloc_traits.h:393
#8  0x000000000062d977 in std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> (this=0xffffa0000d2b6400, __a=...)
    at /usr/include/c++/4.8/bits/shared_ptr_base.h:399
#9  0x000000000062d6e2 in __gnu_cxx::new_allocator<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2> >::construct<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2><std::allocator<osv::application> const, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> > (
    this=0x200003effad7, __p=0xffffa0000d2b6400) at /usr/include/c++/4.8/ext/new_allocator.h:120
#10 0x000000000062d383 in std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2> > >::_S_construct<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2><std::allocator<osv::application> const, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> >(std::allocator<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2> >&, std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2> > >::__construct_helper*, (std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2><std::allocator<osv::application> const, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&>&&)...) (__a=..., __p=0xffffa0000d2b6400) at /usr/include/c++/4.8/bits/alloc_traits.h:254
#11 0x000000000062cd8c in std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2> > >::construct<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2><std::allocator<osv::application> const, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> >(std::allocator<std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2> >&, std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2><s---Type <return> to continue, or q <return> to quit---
td::allocator<osv::application> const, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&>*, (std::_Sp_counted_ptr_inplace<osv::application, std::allocator<osv::application>, (__gnu_cxx::_Lock_policy)2><std::allocator<osv::application> const, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&>&&)...) (
    __a=..., __p=0xffffa0000d2b6400) at /usr/include/c++/4.8/bits/alloc_traits.h:393
#12 0x000000000062c80a in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<osv::application, std::allocator<osv::application>, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> (this=0x200003effe28, __a=...)
    at /usr/include/c++/4.8/bits/shared_ptr_base.h:502
#13 0x000000000062c531 in std::__shared_ptr<osv::application, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<osv::application>, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> (this=0x200003effe20, __tag=..., __a=...)
    at /usr/include/c++/4.8/bits/shared_ptr_base.h:957
#14 0x000000000062c070 in std::shared_ptr<osv::application>::shared_ptr<std::allocator<osv::application>, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> (this=0x200003effe20, __tag=..., __a=...) at /usr/include/c++/4.8/bits/shared_ptr.h:316
#15 0x000000000062b6c4 in std::allocate_shared<osv::application, std::allocator<osv::application>, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&>
    (__a=...) at /usr/include/c++/4.8/bits/shared_ptr.h:598
#16 0x000000000062a902 in std::make_shared<osv::application, std::string const&, std::vector<std::string, std::allocator<std::string> > const&, bool&, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > > const*&> ()
    at /usr/include/c++/4.8/bits/shared_ptr.h:614
#17 0x000000000062583b in osv::application::run (command="/usr/lib/mpi_hello.so", 
    args=std::vector of length 4, capacity 4 = {...}, new_program=true, env=0xffffa0000d2bcc40) at core/app.cc:128
#18 0x000000000060b910 in osv::run (path="/usr/lib/mpi_hello.so", args=std::vector of length 4, capacity 4 = {...}, 
    return_code=0x200003effe0c, new_program=true, env=0xffffa0000d2bcc40) at core/run.cc:13
#19 0x000000000060bbe4 in osv_thread_run_app_in_namespace (filename=0xffffa000058039c0 "/usr/lib/mpi_hello.so", 
    args=0xffffa0000b8936a0, envp=0xffffa0000d2bcc40, thread_id=0x2000004fe488, notification_fd=11) at core/run.cc:64
#20 0x0000000000610c5a in std::_Bind_simple<int (*(char const*, std::vector<std::string, std::allocator<std::string> >*, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >*, long*, int))(char const*, std::vector<std::string, std::allocator<std::string> > const*, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >*, long*, int)>::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul>(std::_Index_tuple<0ul, 1ul, 2ul, 3ul, 4ul>) (this=0xffffa0000d2f16b0) at /usr/include/c++/4.8/functional:1732
#21 0x0000000000610a3f in std::_Bind_simple<int (*(char const*, std::vector<std::string, std::allocator<std::string> >*, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >*, long*, int))(char const*, std::vector<std::string, std::allocator<std::string> > const*, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >*, long*, int)>::operator()() (this=0xffffa0000d2f16b0)
    at /usr/include/c++/4.8/functional:1720
#22 0x0000000000610950 in std::thread::_Impl<std::_Bind_simple<int (*(char const*, std::vector<std::string, std::allocator<std::string> >*, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >*, long*, int))(char const*, std::vector<std::string, std::allocator<std::string> > const*, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >*, long*, int)> >::_M_run() (
    this=0xffffa0000d2f1698) at /usr/include/c++/4.8/thread:115
#23 0x000000000074dfe0 in execute_native_thread_routine ()
#24 0x0000000000681daf in pthread_private::pthread::pthread(void* (*)(void*), void*, sigset_t, pthread_private::thread_attr const*)::{lambda()#1}::operator()() const () at libc/pthread.cc:101
#25 0x00000000006843e6 in std::_Function_handler<void(), pthread_private::pthread::pthread(void* (*)(void*), void*, sigset_t, const pthread_private::thread_attr*)::__lambda6>::_M_invoke(const std::_Any_data &) (__functor=...)
    at /usr/include/c++/4.8/functional:2071
#26 0x000000000044c5e2 in std::function<void ()>::operator()() const (this=0xffff80000d47e080)
    at /usr/include/c++/4.8/functional:2471
#27 0x00000000005bc8aa in sched::thread::main (this=0xffff80000d47e050) at core/sched.cc:1051
#28 0x00000000005b9045 in sched::thread_main_c (t=0xffff80000d47e050) at arch/x64/arch-switch.hh:164
#29 0x000000000048b856 in thread_main () at arch/x64/entry.S:113

(gdb) print &balloon_shrinker
$5 = (jvm_balloon_shrinker **) 0x16a41e0 <balloon_shrinker>
(gdb) print balloon_shrinker
$6 = (jvm_balloon_shrinker *) 0xffffa0000d3f5600
```

(here I restarted VM because mpirun timedout, to get the second part. But addresses should be same, I hope) Second part is during app termination.

```
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 222219264
New value = 222164480
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa00009442e00, obj=0xffffa0000953ac00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 222164480
New value = 222166016
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa00009442e00, obj=0xffffa00009530400)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 222166016
New value = 217273344
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000cca7e00, obj=0xffffa0000c8f5400)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 217273344
New value = 228774912
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000cca7e00, obj=0xffffa0000c8f5a00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 228774912
New value = 228776448
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000c029e00, obj=0xffffa0000c124c00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 228776448
New value = 228713472
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000b9ace00, obj=0xffffa0000bae6c00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 228713472
New value = 228715008
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000a0b4e00, obj=0xffffa0000a1ecc00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 228715008
New value = 23740976
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa00007af9a00, obj=0xffffa00007c30800)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 23740976
New value = 223787008
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000c029e00, obj=0xffffa0000c11a400)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 223787008
New value = 234710016
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000b9ace00, obj=0xffffa0000badc400)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 234710016
New value = 234910720
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa00009442e00, obj=0xffffa00009530a00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 234910720
New value = 10605168
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000b0dae00, obj=0xffffa0000b49d400)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 10605168
New value = 10604814
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000a0b4e00, obj=0xffffa0000a1e1400)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 10604814
New value = 10604824
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa00007af9a00, obj=0xffffa00007c30e00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 10604824
New value = 234912256
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa00009442e00, obj=0xffffa0000953a600)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 234912256
New value = 234869248
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000d927c00, obj=0xffffa0000da1ea00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 234869248
New value = 234872320
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000c029e00, obj=0xffffa0000c11aa00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 234872320
New value = 241460736
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000b9ace00, obj=0xffffa0000badca00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 241460736
New value = 241422848
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa00009a79e00, obj=0xffffa00009ba8a00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 241422848
New value = 241425920
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000b0dae00, obj=0xffffa0000b49da00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 241425920
New value = 1
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000a0b4e00, obj=0xffffa0000a1e1a00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 1
New value = 10605200
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000b9ace00, obj=0xffffa0000bae6600)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 10605200
New value = 10604854
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa0000cca7e00, obj=0xffffa0000cda1600)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 10604854
New value = 23741064
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa00006839e00, obj=0xffffa00006972c00)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
Continuing.
Hardware watchpoint 1: *0x16a41e0

Old value = 23741064
New value = 0
0x0000000000404d40 in elf::program::del_debugger_obj (this=0xffffa000061fca00, obj=0xffffa00006300800)
    at core/elf.cc:1250
1250	    while ((p[0] = p[1]) != nullptr) {
(gdb) c
```

**Comment by justinc1** [comment]
Added some dbg printf, and I actually try to use more than 100 slots from object* program::s_objs[100]; . So add_debugger_obj and del_debugger_obj are broken, as they both traverse beyond end of array. Part of nm loader.elf agrees - balloon_shrinker close after s_objs:
```
00000000016a3c20 B _ZN3elf7program6s_objsE
00000000016a3f40 b _ZStL8__ioinit
00000000016a3f60 b _ZN3mmuL15trace_clear_pteE
00000000016a3fe0 b trace_elf_load
00000000016a4060 b trace_elf_unload
00000000016a40e0 b trace_elf_lookup
00000000016a4160 b trace_elf_lookup_addr
00000000016a41d0 b _ZZ9getauxvalE2_x
00000000016a41e0 B balloon_shrinker
```

Should std::array be used? Any chance of concurrent access - is locking required?

Case when OSv httpserver or cli.so loads app1 in new elf namspace, and app1 loads app2 in another new ELF namespace would be more difficult (and openmpi doesn't do such thing, so we don't need it). App1 would see s_objs at different memory address than httpserver/cli.so anyway, so they couldn't corrupt each other s_objs - they just would not be aware of each other.

PS: Question - does ELF namespace usage means each lib is loaded N-times, consuming N slots?

**Comment by nyh** [comment]
Jusin, thanks, all your comments are right on the money, and I think I should be able to provide a patch fairly easily (I'll do this later) or you can work around this bug temporarily by just setting the array size to 10,000 ;-)

The purpose of this "debugger object" array was to let our gdb "osv syms" code (see osv_syms() in scripts/loader.py) easily find a list of objects we need to load into gdb. We made it a simple C array to make it easy for the script to iterate over it (see commit fa0826a1), but to be honest, I think we could have just used an std::vector (not std::array) or std::list, and it wouldn't have been much harder to process it: we already process STL and Boost data structures several places in loader.py, we even have a function read_vector(). If we must use a fixed-sized array (and I think we don't...) we definitely need to check we didn't overflow it!! But I think it would be easiest (make most C++-like) to switch to std::vector or std::list.

About your questions:

1. Yes, if the same library is loaded several times in different namespaces, each is loaded into a different address, has different global variables, etc., so gdb does need to be aware of them separately, so the same object might appear multiple times in s_objs. Whether this actually works in Gdb, I never actually tested - I hope it does.

2. You're right about the need for locking: program::get_library() does lock the program's "_mutex" throughout the whole operation, so other things are save, but s_objs specifically is shared between all programs (ELF namespaces) so we need another lock to protect it. I don't think this caused any problems you saw, but it needs fixing too.

3. Note that "s_objs" is a static member, i.e., all the instances of program see the same one. There is not a separate one per ELF namespace.

4. I don't understand how 'scripts/build mode=debug image=cli' used 392 objects.  I need to see what it uses on my system, and why. If it uses 392 objects, then "osv syms" would have loaded 392 objects taking ages - and I don't remember this is the case... Weird.

**Comment by justinc1** [comment]
1.
The gdb is currently only partially aware of libs in ELF namespaces - I think if the libmpi.so was loaded in 3 namespaces, only symbols in one of them were resolved (or was that when using backtrace() from within code? I can't remember for sure).

4.
The 392 number - elf.cc as
```
object* program::s_objs[1000];
int obj_ii_max = 0;

void program::add_debugger_obj(object* obj)
{
    auto p = s_objs;
    int ii=0;
    while (*p) {
        ++p;
        ++ii;
        if(ii>=1000) {
            fprintf(stderr, "elf.cc:%d BUG add_debugger_obj arr overflow\n");
            abort();
        }
    }
    if(obj_ii_max>ii) {
        fprintf(stderr, "elf.cc:%d INFO add_debugger_obj obj_ii_max=%d\n", obj_ii_max);
    }
    obj_ii_max = max(ii, obj_ii_max);
    *p = obj;
}
```
And that large number was printed during build.

**Comment by nyh** [comment]
Thanks, I'll send patches, to use std::vector, later tonight.

By the way, for the default build (cli image), the number of objects on this list is 27 (so as you noticed, the "341" number was just a printing bug). I was surprised by the large number - apparently a lot of libraries are needed for Lua. But it didn't overflow the default array size (100) which is why we haven't noticed this bug much earlier.

**Comment by nyh** [comment]
Hi @justinc1, can you please verify that my patch for this issue (already committed) fixes the "4 in 100" crash you saw?
Hope this will bring down the statistic to 0 in 100 crash, but please let me know if not...
Thanks for finding all these bugs!

**Comment by justinc1** [comment]
Yes , it is fixed now. Thank you, Nadav.
